src/meta/processors/partsMan/CreateSpaceProcessor.cpp

```diff
@@ -174,34 +200,97 @@ CreateSpaceProcessor::pickHosts(PartitionID partId,
     return pickedHosts;
 }
 
-std::vector<HostAddr>
-CreateSpaceProcessor::pickHostsWithZone(PartitionID partId,
-                                        const std::vector<std::string>& zones,
-                                        int32_t replicaFactor) {
-    auto startIndex = partId;
-    std::vector<HostAddr> pickedHosts;
-    for (int32_t i = 0; i < replicaFactor; i++) {
-        auto zoneName = zones[startIndex++ % zones.size()];
-        auto zoneIdRet = getZoneId(zoneName);
-        if (!zoneIdRet.ok()) {
-            LOG(ERROR) << "Zone " << zoneName << " not found";
-            handleErrorCode(cpp2::ErrorCode::E_NOT_FOUND);
-            onFinished();
-            return pickedHosts;
+StatusOr<std::unordered_map<HostAddr, int32_t>>
+CreateSpaceProcessor::getHostLoading() {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix();
+    auto code = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return Status::Error("List Parts Failed");
+    }
+
+    std::unordered_map<HostAddr, int32_t> result;
+    while (iter->valid()) {
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto& host : hosts) {
+            result[host]++;
         }
+        iter->next();
+    }
+    return result;
+}
 
+StatusOr<std::vector<HostAddr>>
+CreateSpaceProcessor::pickHostsWithZone(const std::vector<std::string>& zones,
+                                        std::unordered_map<HostAddr, int32_t>& loading) {
+    std::vector<HostAddr> pickedHosts;
+    for (auto& zoneName : zones) {
         auto zoneKey = MetaServiceUtils::zoneKey(zoneName);
         auto zoneValueRet = doGet(std::move(zoneKey));
+        auto hosts = MetaServiceUtils::parseZoneHosts(std::move(zoneValueRet).value());
```

Comment:
No need to read so many times.

Reply:
Do you think it's really necessary? Twice I can take it.